src/main/java/com/microsoft/jenkins/kubernetes/ResourceManager.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.microsoft.jenkins.kubernetes;
+
+import com.microsoft.jenkins.kubernetes.util.CommonUtils;
+import com.microsoft.jenkins.kubernetes.util.Constants;
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.models.V1ObjectMeta;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+public abstract class ResourceManager {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    public static final String DEFAULT_PRETTY = "true";
+
+    protected abstract class ResourceUpdater<T> {
+        private final T resource;
+        private final V1ObjectMeta metadata;
+
+        ResourceUpdater(T resource) {
+            checkNotNull(resource);
+            this.resource = resource;
+            V1ObjectMeta meta = null;
+            try {
+                Method method = resource.getClass().getMethod("getMetadata");
+                meta = (V1ObjectMeta) method.invoke(resource);
+            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+                logger.error(String.format("Fail to fetch meta data for %s", resource));
+            }
+            metadata = meta;
+            checkState(StringUtils.isNotBlank(getName()),
+                    Messages.KubernetesClientWrapper_noName(), getKind(), resource);
+        }
+
+        final String getNamespace() {
+            if (metadata != null) {
+                return metadata.getNamespace();
+            }
+            return Constants.DEFAULT_KUBERNETES_NAMESPACE;
+        }
+
+        final T get() {
+            return resource;
+        }
+
+        final String getName() {
+            String name = null;
+            if (metadata != null) {
+                name = metadata.getName();
+            }
+            return name;
+        }
+
+        @Deprecated
+        final String getKind() {
+            return "GetKind not support in this version";
```

Comment:
Throw an Exception? And move this to Messages.properties?

Reply:
Since the new sdk has kind info in their class name, I think we can use `resource.getClass().getSimpleName()` to get this info. FYI, this method only used in log info in our code.